From 4fc9bc4bedf7911ca60840bf7f1278f565e9c9d4 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Thu, 22 Dec 2016 11:56:44 +0100
Subject: [PATCH 0672/1345] telephony: mvebu: rework tdm2c stop tasklet
 callback

commit  e56d92cdee76e96668e97a2507dc65cea07a6b8b from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

* Rename to "reset" according to actual functionality.
* Remove magics and improve style.

Change-Id: Icd83a569e1ffe45728d18cc27cee7090b6150baf
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/35041
Reviewed-by: Omri Itach <omrii@marvell.com>
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/telephony/mvebu_phone/mv_phone_dev.c |   33 ++++++++++++++------------
 drivers/telephony/mvebu_phone/tdm2c/tdm2c.h  |    2 ++
 2 files changed, 20 insertions(+), 15 deletions(-)

diff --git a/drivers/telephony/mvebu_phone/mv_phone_dev.c b/drivers/telephony/mvebu_phone/mv_phone_dev.c
index 76e6d23..5defc2b 100644
--- a/drivers/telephony/mvebu_phone/mv_phone_dev.c
+++ b/drivers/telephony/mvebu_phone/mv_phone_dev.c
@@ -122,16 +122,14 @@
 static void tdmmc_if_pcm_rx_process(unsigned long arg);
 static void tdm2c_if_pcm_tx_process(unsigned long arg);
 static void tdmmc_if_pcm_tx_process(unsigned long arg);
-
-/* TDM SW Reset */
-static void tdm2c_if_stop_channels(unsigned long args);
+static void tdm2c_if_reset_channels(unsigned long arg);
 
 /* Globals */
 static DECLARE_TASKLET(tdm2c_if_rx_tasklet, tdm2c_if_pcm_rx_process, 0);
 static DECLARE_TASKLET(tdmmc_if_rx_tasklet, tdmmc_if_pcm_rx_process, 0);
 static DECLARE_TASKLET(tdm2c_if_tx_tasklet, tdm2c_if_pcm_tx_process, 0);
 static DECLARE_TASKLET(tdmmc_if_tx_tasklet, tdmmc_if_pcm_tx_process, 0);
-static DECLARE_TASKLET(tdm2c_if_stop_tasklet, tdm2c_if_stop_channels, 0);
+static DECLARE_TASKLET(tdm2c_if_reset_tasklet, tdm2c_if_reset_channels, 0);
 static DEFINE_SPINLOCK(tdm_if_lock);
 static u8 *rx_buff, *tx_buff;
 static u32 rx_miss, tx_miss;
@@ -540,7 +538,7 @@ static irqreturn_t tdm_if_isr(int irq, void *dev_id)
 			dev_dbg(priv->dev, "Stopping the TDM\n");
 			tdm2c_if_pcm_stop();
 			pcm_stop_flag = 0;
-			tasklet_hi_schedule(&tdm2c_if_stop_tasklet);
+			tasklet_hi_schedule(&tdm2c_if_reset_tasklet);
 		} else {
 			dev_dbg(priv->dev, "Tasklet already runningstop_flag\n");
 			pcm_stop_flag = 1;
@@ -637,7 +635,7 @@ static void tdm2c_if_pcm_rx_process(unsigned long arg)
 		spin_lock_irqsave(&tdm_if_lock, flags);
 		pcm_stop_flag = 0;
 		spin_unlock_irqrestore(&tdm_if_lock, flags);
-		tasklet_hi_schedule(&tdm2c_if_stop_tasklet);
+		tasklet_hi_schedule(&tdm2c_if_reset_tasklet);
 	}
 }
 
@@ -696,7 +694,7 @@ static void tdm2c_if_pcm_tx_process(unsigned long arg)
 		spin_lock_irqsave(&tdm_if_lock, flags);
 		pcm_stop_flag = 0;
 		spin_unlock_irqrestore(&tdm_if_lock, flags);
-		tasklet_hi_schedule(&tdm2c_if_stop_tasklet);
+		tasklet_hi_schedule(&tdm2c_if_reset_tasklet);
 	}
 }
 
@@ -725,24 +723,27 @@ static void tdmmc_if_pcm_tx_process(unsigned long arg)
 	spin_unlock_irqrestore(&tdm_if_lock, flags);
 }
 
-static void tdm2c_if_stop_channels(unsigned long arg)
+/* TDM2C restart channel callback */
+static void tdm2c_if_reset_channels(unsigned long arg)
 {
 	u32 max_poll = 0;
 	unsigned long flags;
-	void __iomem *tdm_base = get_tdm_base();
 
 	/* Wait for all channels to stop  */
-	while (((readl(tdm_base + CH_ENABLE_REG(0)) & 0x101) ||
-		(readl(tdm_base + CH_ENABLE_REG(1)) & 0x101)) && (max_poll < 30)) {
+	while (((readl(priv->tdm_base + CH_ENABLE_REG(0)) & CH_RXTX_EN_MASK) ||
+		(readl(priv->tdm_base + CH_ENABLE_REG(1)) & CH_RXTX_EN_MASK)) &&
+		(max_poll < MV_TDM_STOP_POLLING_TIMEOUT)) {
+
 		mdelay(1);
 		max_poll++;
 	}
 
 	dev_dbg(priv->dev, "Finished polling on channels disable\n");
-	if (max_poll >= 30) {
-		writel(0, tdm_base + CH_ENABLE_REG(0));
-		writel(0, tdm_base + CH_ENABLE_REG(1));
-		dev_warn(priv->dev, "\n\npolling on channels disabling exceeded 30ms\n\n");
+	if (max_poll >= MV_TDM_STOP_POLLING_TIMEOUT) {
+		writel(0, priv->tdm_base + CH_ENABLE_REG(0));
+		writel(0, priv->tdm_base + CH_ENABLE_REG(1));
+		dev_warn(priv->dev, "\n%s: Channels disabling timeout (%dms)\n",
+			 __func__, MV_TDM_STOP_POLLING_TIMEOUT);
 #ifdef CONFIG_MV_TDM_EXT_STATS
 		pcm_stop_fail++;
 #endif
@@ -752,6 +753,8 @@ static void tdm2c_if_stop_channels(unsigned long arg)
 	spin_lock_irqsave(&tdm_if_lock, flags);
 	is_pcm_stopping = 0;
 	spin_unlock_irqrestore(&tdm_if_lock, flags);
+
+	/* Restart channels */
 	tdm2c_if_pcm_start();
 }
 
diff --git a/drivers/telephony/mvebu_phone/tdm2c/tdm2c.h b/drivers/telephony/mvebu_phone/tdm2c/tdm2c.h
index 70d7fdc..b124b51 100644
--- a/drivers/telephony/mvebu_phone/tdm2c/tdm2c.h
+++ b/drivers/telephony/mvebu_phone/tdm2c/tdm2c.h
@@ -126,6 +126,7 @@
 #define CH_TX_CUR_ADDR_REG(ch)	(0x60 | ((ch) << 3))
 #define CH_RX_CUR_ADDR_REG(ch)	(0x64 | ((ch) << 3))
 #define CH_ENABLE_REG(ch)	(((ch) + 1) << 4)
+#define CH_RXTX_EN_MASK		0x101
 #define CH_BUFF_OWN_REG(ch)	(0x04 | (((ch) + 1) << 4))
 #define CH_TX_ADDR_REG(ch)	(0x08 | (((ch) + 1) << 4))
 #define CH_RX_ADDR_REG(ch)	(0x0c | (((ch) + 1) << 4))
@@ -254,6 +255,7 @@
 #define MV_TDM_MAX_SAMPLING_PERIOD		30	/* ms */
 #define MV_TDM_BASE_SAMPLING_PERIOD		10	/* ms */
 #define MV_TDM_TOTAL_CH_SAMPLES			80	/* samples */
+#define MV_TDM_STOP_POLLING_TIMEOUT		30	/* ms */
 
 /* TDM IRQ types */
 #define MV_EMPTY_INT		0
-- 
1.7.9.5

